compiler/rustc_parse/src/parser/path.rs

```diff
@@ -435,38 +436,92 @@ impl<'a> Parser<'a> {
             if let AssocTyConstraintKind::Bound { .. } = kind {
                 self.sess.gated_spans.gate(sym::associated_type_bounds, span);
             }
-
-            let constraint = AssocTyConstraint { id: ast::DUMMY_NODE_ID, ident, kind, span };
+            let constraint =
+                AssocTyConstraint { id: ast::DUMMY_NODE_ID, ident, gen_args: vec![], kind, span };
             Ok(Some(AngleBracketedArg::Constraint(constraint)))
         } else {
-            Ok(self.parse_generic_arg()?.map(AngleBracketedArg::Arg))
+            let lo = self.token.span;
+            let arg = self.parse_generic_arg()?;
+            if self.eat(&token::Eq) {
+                // Parse the type of the equality constraint even if parsing of the associated type failed
+                match self.get_assoc_type_with_generic_args(arg, lo) {
```

Comment:
This logic will subsume the `if self.check_ident() && ...` branch above, but you need to handle the `token::Colon` case as well in addition to the `token::Eq` case.